inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,158 @@
+
+namespace vpu {
+
+bool MergeGatherGatherElements::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    bool wasGraphChanged = false;
+
+    const auto gatherData = ngraph::pattern::any_input();
+    const auto gatherIndices = ngraph::pattern::any_input(ngraph::pattern::has_static_rank());
+    const auto gatherAxis = ngraph::pattern::wrap_type<ngraph::opset6::Constant>();
+    const auto gather = ngraph::pattern::wrap_type<ngraph::opset6::Gather>(
+        {gatherData, gatherIndices, gatherAxis},
+        ngraph::pattern::consumers_count(1));
+
+    const auto squeezeAxes = ngraph::pattern::wrap_type<ngraph::opset6::Constant>();
+    const auto squeeze = ngraph::pattern::wrap_type<ngraph::opset6::Squeeze>({gather, squeezeAxes}, ngraph::pattern::consumers_count(1));
+
+    const auto transposePerm = ngraph::pattern::wrap_type<ngraph::opset6::Constant>();
+    const auto transpose = ngraph::pattern::wrap_type<ngraph::opset6::Transpose>({squeeze, transposePerm});
+
+    const auto m = std::make_shared<ngraph::pattern::Matcher>(transpose, "GatherSqueezeTransposeMatcher");
+    for (const auto& node : f->get_ordered_ops()) {
+        if (!m->match(node)) {
+            continue;
+        }
+
+        const auto& patternMap = m->get_pattern_value_map();
+
+        std::vector<ngraph::Node*> shapeOfs;
+        std::vector<ngraph::Node*> gatherElements;
+
+        const auto& matchedTranspose = patternMap.at(transpose);
+        const auto& transposeConsumers = matchedTranspose.get_target_inputs();
+
+        const auto isShapeOfOrGatherElements = [](const ngraph::Input<ngraph::Node>& consumer) {
+            return consumer.get_node()->get_type_info().is_castable(ngraph::opset6::ShapeOf::type_info) ||
+                   consumer.get_node()->get_type_info().is_castable(ngraph::opset6::GatherElements::type_info);
+        };
+
+        if (!std::all_of(transposeConsumers.cbegin(), transposeConsumers.cend(), isShapeOfOrGatherElements)) {
+            continue;
+        }
+
+        for (const auto& transposeConsumer : transposeConsumers) {
+            if (transposeConsumer.get_node()->get_type_info().is_castable(ngraph::opset6::ShapeOf::type_info)) {
+                shapeOfs.push_back(transposeConsumer.get_node());
+            } else if (transposeConsumer.get_node()->get_type_info().is_castable(ngraph::opset6::GatherElements::type_info)) {
+                gatherElements.push_back(transposeConsumer.get_node());
+            }
+        }
+
+        const auto& matchedTransposePerm = patternMap.at(transposePerm);
+        const auto& matchedSqueezeAxes = patternMap.at(squeezeAxes);
+        const auto& matchedSqueeze = patternMap.at(squeeze);
+        const auto& matchedGatherData = patternMap.at(gatherData);
+        const auto& matchedGatherIndices = patternMap.at(gatherIndices);
+        const auto& matchedGatherAxis = patternMap.at(gatherAxis);
+
+        const auto axisConst = ngraph::get_constant_from_source(matchedGatherAxis);
```

Comment:
You need to check that `axisConst` is not nullptr.

Reply:
Discussed offline, will be done in a separate PR.